From f9007c44c7f2e446ff931cb1fd0667fc439902e5 Mon Sep 17 00:00:00 2001
From: chdugw_auto <jkulkarni@maxlinear.com>
Date: Wed, 25 Aug 2021 14:52:58 +0300
Subject: [PATCH] Merge pull request #3484 in SW_UGW/linux from
 feature/PONRTSYS-9417-proper-way-of-handling-queue-map-replication to master

* commit '641365cca55f0932fed062c4b253003d53b1d4b1':
  PONRTSYS-9417: Proper way of handling queue map replication
---
 drivers/net/datapath/pon_qos/pon_qos_tc_qos.c | 62 +++++++++++++++------------
 1 file changed, 34 insertions(+), 28 deletions(-)

diff --git a/drivers/net/datapath/pon_qos/pon_qos_tc_qos.c b/drivers/net/datapath/pon_qos/pon_qos_tc_qos.c
index ae5bb2a531aa..a2a4c57bfdcf 100644
--- a/drivers/net/datapath/pon_qos/pon_qos_tc_qos.c
+++ b/drivers/net/datapath/pon_qos/pon_qos_tc_qos.c
@@ -964,27 +964,14 @@ static int pon_qos_map_qid(struct pon_qos_q_data *qid,
 	return ret;
 }
 
-/* 
- * We need this increment to calculate "next" subif
- * For example, if a subif is 0x200, then the next
- * subif will be 0x300, and next 0x400 and so on
- */
-#define SUBIF_INCREMENT 0x100
-
 /* We need this function so that the command:
  *
  *    tc filter add dev eth0_0 ingress pref 1 protocol 802.1q flower \
- *        skip_sw vlan_prio 0 classid 8002:1 acion ok
- *
- * would trigger mappings for each subif:
+ *        skip_sw vlan_prio 0 classid 8002:1 action ok
  *
- *    subif=0 tc=tc
- *    subif=1 tc=tc
- *    subif=2 tc=tc
- *    subif=3 tc=tc
- *
- * which effectively means that it will "replicate" the mappings for
- * eth0_0 subinterfaces (like: eth0_0_1, eth0_0_2, ...)
+ * would trigger mappings for each subif, which for example effectively means
+ * that it will "replicate" the mappings for eth0_0 subinterfaces (like:
+ * eth0_0_1, eth0_0_2, ...)
  */
 static int pon_qos_replicate_mappings_for_subifs(struct net_device *dev,
 						 struct pon_qos_q_data *qid,
@@ -992,27 +979,46 @@ static int pon_qos_replicate_mappings_for_subifs(struct net_device *dev,
 						 int tc, bool en)
 {
 	struct dp_port_prop prop;
-	int i;
-	int ret;
+	int i = 0;
+	int ret = 0;
 	dp_subif_t subif = {0};
+	struct net_device *netdev;
 
-	if (dp_get_netif_subifid(dev, NULL, NULL, NULL, &subif, 0))
-		return -EINVAL;
-
-	/* Read the number of available sub interfaces */
 	if (dp_get_port_prop(sch->inst, sch->port, &prop))
 		return -EINVAL;
 
+	read_lock(&dev_base_lock);
+
 	/* Perform (subif, tc) -> (queue) mappings for each sub interface */
-	for (i = 0; i < prop.num_subif; ++i) {
-		int subif_id = subif.subif + i * SUBIF_INCREMENT;
+	for_each_netdev(&init_net, netdev) {
+		if (dp_get_netif_subifid(netdev, NULL, NULL, NULL, &subif, 0))
+			continue;
+		if (subif.port_id != sch->port)
+			continue;
 
-		ret = pon_qos_set_qmap(qid, sch->port, subif_id, tc, en);
+		ret = pon_qos_set_qmap(qid, sch->port, subif.subif, tc, en);
 		if (ret)
-			return -EINVAL;
+			goto cleanup;
+		i++;
 	}
 
-	return 0;
+cleanup:
+	if (ret) {
+		for_each_netdev(&init_net, netdev) {
+			if (dp_get_netif_subifid(netdev, NULL, NULL, NULL, &subif, 0))
+				continue;
+			if (subif.port_id != sch->port)
+				continue;
+			if (i <= 0)
+				break;
+
+			pon_qos_set_qmap(qid, sch->port, subif.subif, tc, !en);
+			i--;
+		}
+	}
+
+	read_unlock(&dev_base_lock);
+	return ret;
 }
 
 /* classid option identifies the queue id.
